A visitor for exceeding canvas (and margin) by a complex command

jowczar opened the pull request.

jowczar committed:
Fix not compiling code

jowczar committed:
Add command exceeding basic canvas

jowczar committed:
Merge branch 'master' into feature/exceeding-canvas-visitor

Artur229322 committed:
Created visitor class and cleaned up CommandsCounterVisitor

Artur229322 committed:
Merge remote-tracking branch 'origin/feature/exceeding-canvas-visitor' into feature/exceeding-canvas-visitor

jowczar committed:
Add dummy CommandBoundariesCheckVisitor methods content

jowczar committed:
Revert "Add dummy CommandBoundariesCheckVisitor methods content"

This reverts commit 09e337e35266db7fb7be781241b0c3ff48db1bc5.

jowczar committed:
Add Subscriber for CommandBoundariesCheckVisitor

Artur229322 committed:
Finished visitor implementation and added required getters

Artur229322 committed:
Added new method

jowczar committed:
Merge remote-tracking branch 'origin/feature/exceeding-canvas-visitor' into feature/exceeding-canvas-visitor

jowczar committed:
Pass canvas to CommandBoundariesCheckVisitor

Artur229322 committed:
Added dimensions checks in visitcommands methods.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java:
```diff
@@ -0,0 +1,45 @@
+import static java.lang.Math.abs;
+
+public class CommandBoundariesCheckVisitor implements Visitor{
+	private final JPanel canvas;
```
Add CanvasChecker abstraction and JPanel adapter.

jowczar committed:
Add ICanvas interface

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java:
```diff
@@ -0,0 +1,43 @@
+	@Override
+	public void visitOperateToCommand(OperateToCommand operateToCommand) {
+		if(abs(operateToCommand.getPosX()) > canvas.getWidth()/2 || abs(operateToCommand.getPosY()) > canvas.getHeight()/2){
+			isExceedingBoundaries = true;
```
Such implementation of the visitor is not reusable in run-time. Once flag is set it cannot be unset and following accept calls would always return true. 
To fix this you should set the flag properly (to true or false!) in each visit method (also in visitICompoundCommand)

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java:
```diff
@@ -0,0 +1,43 @@
+		while(iterator.hasNext()) {
+			DriverCommand driverCommand = iterator.next();
+
+			driverCommand.accept(this);
```
notice that you can short-cut (break) the visitation if isExceedingBoundaries is set to true.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/ICanvas.java:
```diff
@@ -0,0 +1,6 @@
+package edu.kis.powp.jobs2d.command;
+
+public interface ICanvas {
+	 int getWidth();
+	 int getHeight();
```
Please, do not assume that each canvas wound be rectangular (you can change name to IRectangularCanvas)

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java:
```diff
@@ -0,0 +1,43 @@
+
+	@Override
+	public void visitSetPositionCommand(SetPositionCommand setPositionCommand) {
+		if(abs(setPositionCommand.getPosX()) > canvas.getWidth()/2 || abs(setPositionCommand.getPosY()) > canvas.getHeight()/2){
```
Code duplication. You should apply SRP: checking this condition can be delegated to some strategy.

kowallus reviewed (changes requested):
This solution can be improved in a few aspects.

Artur229322 committed:
Cleaned up visitor functions and added strategy interface.

Artur229322 committed:
Changed implementation to use method from strategy instead.

Artur229322 committed:
Changed checkExceedingBoundaries parameters to X and Y

Artur229322 committed:
Changed checkExceedingBoundaries parameters to X and Y and renamed ICanvas

Artur229322 committed:
Merge branch 'master' into feature/exceeding-canvas-visitor

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java:
```diff
@@ -4,11 +4,11 @@
-public class CommandBoundariesCheckVisitor implements Visitor{
-	private final ICanvas canvas;
+public class CommandBoundariesCheckVisitor implements Visitor, ICommandBoundariesCheckStrategy{
+	private final IRectangularCanvas canvas;
```
CommandBoundariesCheckVisitor doesn't need to assume work with rectangular canvas. It is an unnecessary limitation.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java:
```diff
@@ -4,11 +4,11 @@
 
-public class CommandBoundariesCheckVisitor implements Visitor{
-	private final ICanvas canvas;
+public class CommandBoundariesCheckVisitor implements Visitor, ICommandBoundariesCheckStrategy{
```
Implementing ICommandBoundariesCheckStrategy is closer to the Template Method pattern approach. In Strategy pattern ICommandBoundariesCheckStrategy would be an attribute. Please refactor.

kowallus reviewed (changes requested):
It is not how the Strategy pattern is implemented. Refactor.

jowczar committed:
Generalize IRectangularCanvas to ICanvas

jowczar committed:
Make check strategy an actual strategy pattern

jowczar committed:
Rename strategy

jowczar committed:
Rm unused import

jowczar committed:
Fix bug in calculations

jowczar commented:
@kowallus
I made a couple of changes.

– I changed it so now check strategy is an actual staregy pattern. It will hold the logic of calculation of canvas, while Visitor only takes care of executing it on every command.

– I also made Visitor accept any canvas as I don't think it should be limited/determined in the visitor

– I deleted IRectangularCanvas as I didn't see further use for it for now. I think all canvas should have width and height (the outermost points in case of shapes) as it may be helpful in future functionalities so I left those getters in ICanvas. 

If they are in a shape of, for example, a star, they can still use RectangularCanvasCheckStrategy to get "box" calculation or somebody can implement FigureCanvasCheckStrategy and calculate exact field of canvas and whether the provided point is out of it. 

If there is a need for any other weird calculation, I think this is the way it is possible to be extended without changes in our code.

– I wasn't sure about naming and placement of the files. I renamed the ICommandBoundairesCheckStrategy to ICanvasBoundariesCheckStrategy. It can be used anywhere, it doesn't have to be a command-related.

– finally, I moved abs() into calculation function.

LMK if anything needs change

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/ICanvasBoundariesCheckStrategy.java:
```diff
@@ -0,0 +1,5 @@
+package edu.kis.powp.jobs2d.command;
+
+public interface ICanvasBoundariesCheckStrategy {
+	boolean checkExceedingBoundaries(ICanvas canvas, int X, int Y);
```
IMO: canvas parameter is not necessary here. ICanvas object can be simply a attribute of ICanvasBoundariesCheckStrategy implementation.

Artur229322 committed:
Moved ICanvas to strategy implementation.

Artur229322 committed:
Changed ICanvas to IRectangularCanvas
